### ISSUE TITLE
Custom context property over subscriptions

### ISSUE THREAD

**Issue by fenos** [Urigo/graphql-modules#694]
Hi there, firstly, thanks a lot for this AWESOME library, really enjoying it so far.
Everything was going too smooth until I've started working on subscriptions.

#### Problem
I'm struggling on having a consistent `authToken` property within the graphql context for regular `Mutation/Queries` and `Subscription`.

#### Let me show you some code

I have my `AuthModule` which nicely inject the `authToken` within the context as following:

```ts
export interface ISession {
    req: express.Request;
    res: express.Response;
    socket?: WebSocket,
}
export const AuthModule = new GraphQLModule({
    imports: [
        DBModule,
    ],
    providers: [
        AuthService,
        AuthRepository,
        AuthEvents,
    ],
    resolvers: loadResolversFiles(path.join(__dirname, "resolvers/")),
    typeDefs: loadSchemaFiles(path.join(__dirname, "*.graphql")),

    async context(session: ISession, currentContext, { injector }) {
        const jwtToken = session.req.headers.authorization;
        const authToken = injector.get(AuthService)
            .createToken(jwtToken ? jwtToken.replace("Bearer ", "") : "")
            .parse();

        return {
            authToken,
        };
    },
});
```

Until here it's all good! I can access the `authToken` from within the **resolver context** from any module that imports this `AuthModule`. Sweet!

#### GraphQL Subscription
Now switching to a **subscription** I'm obviously getting:

```
cannot read property 'headers' of undefined
```

which make sense since the `session.req` is undefined and now I have a `socket` instance!

from the **ApolloServer** documentation I'm able to detect from the context function if the current request is a **connection (subscription)** or a **request (mutation / query)** and return a slightly different context which is generated by the `Subscription server hooks` example:

```js
const server = new ApolloServer({
  schema,
  context: async ({ req, connection }) => {
    if (connection) {
      // check connection for metadata
      return connection.context;
    } else {
      // check from req
      const token = req.headers.authorization || "";

      return { token };
    }
  },
});
```

With **GraphQL modules** I'm unable to decorate the `context` when I'm under a `subsctiption` operation for the reason that I cannot access the `connectionParams` that are visible within the **onConnect** hooks etc... 
The **connection** object also doesn't exist...

I also tried to implement the `onConnect` hook for the subscription and returning the `authToken` as following:

```ts
onConnect(connectionParams: {Authorization?: string}) {
        const authToken = connectionParams.Authorization;

        if (!authToken) {
            throw UnauthorizedError("User Unauthorized");
        }

        const token = this.authService.createToken(
            (authToken).replace("Bearer ", ""),
        ).parse();

        if (!token.isValid()) {
            throw UnauthorizedError("User Unauthorized");
        }

        return {authToken: token};
    }
```

but unfortunately, that returned object hasn't appeared in the `context` function of the module.

Is there a way to achieve what I'm after?

#### Reasons
The reason behind having the `authToken` within the context makes it consistent to have authentication and authorization checks, with the ability to use the `resolver` composition to abstract away the logic. 

At the current stage, it doesn't look that is possible with `subscriptions`

Please advise 👍

**Comment by fenos** [comment]
I'm still investigating but so far seems that the issue is caused by functions calls in the wrong order.

The `context` function is called before any other subscription `hook` thus making it impossible to add properties to the context

**Comment by fenos** [comment]
After a few hours of debugging the only conclusion I came up with was to include the `connectionParams` within the function call `context` when the `onConnect` hook is invoked. 

In that way, we can access the subscription parameters and read the `authorization` token

Super Simple PR above 👆

**Comment by fenos** [comment]
Here is an example that reproduce the issue:

https://codesandbox.io/embed/typescript-node-c4gkl

**Comment by ardatan** [comment]
@fenos Thank you for opening this issue! I added a new option property called `subscriptionHooks` for the people who don't use DI for subscription hooks;
Canary version;`0.7.13-alpha-2deabb5.19+2deabb5`
```
new GraphQLModule({
  subscriptionHooks: {
    onConnect(connectionParams: {Authorization?: string}) {
        const authToken = connectionParams.Authorization;

        if (!authToken) {
            throw UnauthorizedError("User Unauthorized");
        }

        const token = this.authService.createToken(
            (authToken).replace("Bearer ", ""),
        ).parse();

        if (!token.isValid()) {
            throw UnauthorizedError("User Unauthorized");
        }

        return {authToken: token};
    }
  }
})
```

**Comment by fenos** [comment]
@ardatan absolutely appreciated! 
That's a neat way of handling this!

**Comment by ardatan** [comment]
Available in 0.5.13!

**Comment by ArsenAvagyan** [comment]
Context is not beeing changed after returning something from onConnect hook. Do you have some working code example?
